### PR TITLE
Implement Citation and Reference node

### PR DESCRIPTION
**Pull request by InnocentBug**
# Description
Implementation of the Citation and Reference node.

## Changes
I had to move the reference and software node from primary to subobjects.

## Known Issues
None
## Notes

### CODE REVIEW

File: src/cript/nodes/subobjects/reference.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+
+class Reference(BaseNode):
+    """
+    Reference node
+    """
+    @dataclass(frozen=True)
+    class JsonAttributes(BaseNode.JsonAttributes):
+        node: str = "Reference"
+        url: str = ""
```
[the remark this answers is not in the document]
Comment (nh916) [reply src/cript/nodes/subobjects/reference.py:13]: I fully agree with you!

[the remark this answers is not in the document]
Comment (InnocentBug) [reply src/cript/nodes/subobjects/reference.py:13]: I agree and have noticed it too.
Maybe a URL Base Node.

BaseNode -> Subobject
BaseNode -> URLBaseNode
URLBaseNode -> PrimaryBaseNode
URLBaseNode -> Software, Citation
URLBaseNode -> Supporting nodes.

File: src/cript/nodes/core.py [hunk 2]
```diff
@@ -37,4 +47,28 @@ def validate(self) -> None:
...
+        # All attributes from the backend are passed over, but some like created_by are ignored
+        node = cls(**json)
+        # Now we push the full json attributes into the class if it is valid
+        attrs = cls.JsonAttributes(**json)
+        node._update_json_attrs_if_valid(attrs)
+        return node
+
+
+    @property
+    def json(self):
```
[the remark this answers is not in the document]
Comment (nh916) [reply src/cript/nodes/core.py:67]: okay that makes sense

File: src/cript/nodes/subobjects/reference.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nh916) [reply src/cript/nodes/subobjects/reference.py:13]: in the URL base we also want to put UID and UUID's right?

[the remark this answers is not in the document]
Comment (InnocentBug) [reply src/cript/nodes/subobjects/reference.py:15]: yes, that is what @brili said.

**Review by brili — APPROVED** [review]

**Comment by InnocentBug** [comment]
Merged with #28

**Pull request closed without merging**